compiler/z/codegen/S390Peephole.hpp

```diff
@@ -0,0 +1,141 @@
+ * Copyright (c) 2018, 2018 IBM Corp. and others
+ *
+ * This program and the accompanying materials are made available under
+ * the terms of the Eclipse Public License 2.0 which accompanies this
+ * distribution and is available at http://eclipse.org/legal/epl-2.0
+ * or the Apache License, Version 2.0 which accompanies this distribution
+ * and is available at https://www.apache.org/licenses/LICENSE-2.0.
+ *
+ * This Source Code may also be made available under the following Secondary
+ * Licenses when the conditions for such availability set forth in the
+ * Eclipse Public License, v. 2.0 are satisfied: GNU General Public License,
+ * version 2 with the GNU Classpath Exception [1] and GNU General Public
+ * License, version 2 with the OpenJDK Assembly Exception [2].
+ *
+ * [1] https://www.gnu.org/software/classpath/license.html
+ * [2] http://openjdk.java.net/legal/assembly-exception.html
+ *
+ * SPDX-License-Identifier: EPL-2.0 OR Apache-2.0 OR GPL-2.0 WITH Classpath-exception-2.0 OR LicenseRef-GPL-2.0 WITH Assembly-exception
+ *******************************************************************************/
+
+#ifndef OMR_Z_PEEPHOLE
+#define OMR_Z_PEEPHOLE
+
+#include "codegen/CodeGenerator.hpp"
+#include "compile/Compilation.hpp"
+#include "codegen/FrontEnd.hpp"
+#include "codegen/Instruction.hpp"
+#include "env/FilePointerDecl.hpp"
+
+class TR_S390Peephole
+   {
+public:
+   TR_S390Peephole(TR::Compilation* comp, TR::CodeGenerator *cg);
+
+protected:
+   void printInfo(const char* info)
+      {
+      if (_outFile)
+         {
+         if ( !( !comp()->getOption(TR_TraceCG) && comp()->getOptions()->getTraceCGOption(TR_TraceCGPostBinaryEncoding) && comp()->getOptions()->getTraceCGOption(TR_TraceCGMixedModeDisassembly) )  )
+            {
+            trfprintf(_outFile, info);
+            }
+         }
+      }
+
+   void printInst()
+      {
+      if (_outFile)
+         {
+         if ( !( !comp()->getOption(TR_TraceCG) && comp()->getOptions()->getTraceCGOption(TR_TraceCGPostBinaryEncoding) && comp()->getOptions()->getTraceCGOption(TR_TraceCGMixedModeDisassembly) )  )
+            {
+            comp()->getDebug()->print(_outFile, _cursor);
+            }
+         }
+      }
+
+   TR::Compilation * comp() { return TR::comp(); }
+
+protected:
+   TR_FrontEnd * _fe;
+   TR::FILE *_outFile;
+   TR::Instruction *_cursor;
+   TR::CodeGenerator *_cg;
+   };
+
+class TR_S390PreRAPeephole : private TR_S390Peephole
+   {
+public:
+   TR_S390PreRAPeephole(TR::Compilation* comp, TR::CodeGenerator *cg);
+
+   void perform();
+
+private:
+   /** \brief
+    *     Attempts to reduce L[' '|FH|G] R,MR1  ST[' '|FH|G] R,MR2 sequences to MVC MR2, MR1
+    *     to save a register and instruction.
+    *
+    *  \return
+    *     true if the reduction was successful; false otherwise.
+    */
+   bool LoadStoreReduction(TR::InstOpCode::Mnemonic storeOpCode, uint16_t size);
```

Comment:
Although we are pretty inconsistent already what do you think of imposing a convention for these peepholes and naming this one `attemptLoadStoreReduction`? It is an attempt after all and is not guaranteed to succeed. Can we also document the parameters?

Reply:
a456077